src/instructlab/sdg/generate_data.py

```diff
@@ -260,6 +332,169 @@ def load_pipeline(yaml_basename):
     )
 
 
+def _generate_knowledge_qa_dataset(
+    logger, generated_dataset: Dataset, keep_context_separate=False
+):
+    """
+    Generate question and answer pairs from the newly generated dataset
+    for each taxonomy leaf node. Each row of the generated dataset gets
+    converted to have messages, metadata, and id columns.
+
+    If `keep_context_separate` is True, then a context column is also added.
+    If `keep_context_separate` is False, the context colum is omitted and
+    the context is instead added directly to the user message content.
+    """
+
+    def __create_qa_row(rec):
+        msg_id = str(uuid.uuid4())
+        context = rec["document"]
+        instruction = _get_question(logger, rec)
+        response = _get_response(logger, rec)
+        metadata = {
+            "sdg_document": rec["document"],
+            "domain": rec["domain"],
+            "dataset": "document_knowledge_qa",
+        }
+        if "raw_document" in rec and "dataset_type" in rec:
+            metadata.update(
+                {
+                    "raw_document": rec["raw_document"],
+                    "dataset_type": rec["dataset_type"],
+                }
+            )
+        metadata = json.dumps(metadata)
+        if keep_context_separate:
+            messages = [
+                {"role": "user", "content": f"{instruction}"},
+                {"role": "assistant", "content": response},
+            ]
+            return {
+                "messages": messages,
+                "metadata": metadata,
+                "id": msg_id,
+                "context": context,
+            }
+        messages = [
+            {"role": "user", "content": f"{context}\n\n{instruction}"},
+            {"role": "assistant", "content": response},
+        ]
+
+        return {"messages": messages, "metadata": metadata, "id": msg_id}
+
+    knowledge_ds = generated_dataset.map(
+        __create_qa_row, remove_columns=generated_dataset.column_names
+    )
+    return knowledge_ds
+
+
+def _build_raft_dataset(ds: Dataset, p, num_doc_in_context=4):
```

Comment:
ChatGPT offers 2 possible explanations for what "raft" could mean here:

> Random Augmented Feature Training (RAFT): In machine learning, "RAFT" might stand for Random Augmented Feature Training, a technique where random augmentation is applied to the features of a dataset to improve the model's robustness and performance. The function's description suggests that additional context from random samples is added, which could be a form of data augmentation.
>
> Reference Augmented Fine-Tuning (RAFT): In NLP, RAFT might refer to adding additional context or references to enhance the training data. By incorporating context from other samples, the function could be aiming to provide a more comprehensive background, potentially improving the quality of responses generated by a language model.

Seems like it's the latter? Expanding the acronym and linking to https://aka.ms/raft-paper would help

Reply:
I do think this in the spirit of that RAFT paper, but I'll give this a different name that doesn't require understanding the acronym. Something like `_add_extra_contexts_to_samples`  since it's adding extra contexts to each sample in the dataset.